src/vfs2.c

```diff
@@ -1755,27 +1757,46 @@ int vfs2_add_uncommitted(sqlite3_file *file,
 		}
 		sums.cksum1 = ByteGetBe32(prev_fhdr.cksum1);
 		sums.cksum2 = ByteGetBe32(prev_fhdr.cksum2);
+		db_size = ByteGetBe32(prev_fhdr.commit);
 	} else {
+		/* This is the first transaction in this WAL. In this case
+		 * we initialize the rolling checksum from the checksum in
+		 * the WAL header, and read the actual database file to
+		 * initialize the running database size. */
 		sums.cksum1 = ByteGetBe32(e->wal_cur_hdr.cksum1);
 		sums.cksum2 = ByteGetBe32(e->wal_cur_hdr.cksum2);
-	}
+		/* The database size in pages is kept in a field of the database
+		 * header. */
+		uint8_t b[DB_HEADER_SIZE];
+		rv =
+		    xfile->orig->pMethods->xRead(xfile->orig, &b, sizeof(b), 0);
+		/* TODO(cole) this can't fail provided that the main file
+		 * has been created; ensure that this is the case even if
+		 * we haven't run a checkpoint yet. */
+		assert(rv == SQLITE_OK);
+		db_size = ByteGetBe32(b + DB_HEADER_NPAGES_OFFSET);
+	}
+	POST(db_size > 0);
 
 	struct vfs2_shm_region0 *r0 = e->shm_regions[0];
 	PRE(e->wal_cursor < REGION0_PGNOS_LEN);
-	r0->pgnos[e->wal_cursor] = frames[0].page_number;
-	struct wal_frame_hdr fhdr = txn_frame_hdr(e, sums, frames[0]);
-	rv = write_one_frame(e, fhdr, frames[0].page);
+	r0->pgnos[e->wal_cursor] = (uint32_t)frames[0].page_number;
```

Comment:
Don't we have to do a lookup for the page number using the hash?

Reply:
The `pgnos` array just stores the page numbers of each frame in the order they appear in the WAL, and then the hash table stores indices into `pgnos`. So updating pgnos is straightforward. But I need to write a comment here about why we don't update the hash table until later.

Reply:
Added a comment about this.